File: cpp/src/solutions/roman_to_integer.cpp
// LeetCode 13. Roman to Integer

#include "cpp/src/common/common.h"

class Solution {
 public:
  int romanToInt(string s) {
    map<char, int> value_map = {
        {'I', 1},   {'V', 5},   {'X', 10},   {'L', 50},
        {'C', 100}, {'D', 500}, {'M', 1000},
    };

    int ans = 0, last = 0;
    for (auto iter = s.rbegin(); iter != s.rend(); ++iter) {
      auto curr = value_map[*iter];
      if (curr < last) {
        ans -= curr;
      } else {
        ans += curr;
      }
      last = curr;
    }
    return ans;
  }
};

TEST(romanToInt, example1) {
  string s = "III";
  int ans = 3;
  EXPECT_EQ(Solution().romanToInt(s), ans);
}

TEST(romanToInt, example2) {
  string s = "IV";
  int ans = 4;
  EXPECT_EQ(Solution().romanToInt(s), ans);
}

TEST(romanToInt, example3) {
  string s = "IX";
  int ans = 9;
  EXPECT_EQ(Solution().romanToInt(s), ans);
}

TEST(romanToInt, example4) {
  string s = "LVIII";
  int ans = 58;
  EXPECT_EQ(Solution().romanToInt(s), ans);
}

TEST(romanToInt, example5) {
  string s = "MCMXCIV";
  int ans = 1994;
  EXPECT_EQ(Solution().romanToInt(s), ans);
}

File: rust/src/solutions/roman_to_integer.rs
// LeetCode 13. Roman to Integer in Rust

use super::Solution;
use std::collections::HashMap;

impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        let value_map: HashMap<char, i32> = vec![
            ('I', 1),
            ('V', 5),
            ('X', 10),
            ('L', 50),
            ('C', 100),
            ('D', 500),
            ('M', 1000),
        ]
        .into_iter()
        .collect();

        let mut ans = 0;
        let mut last = 0;
        for ch in s.chars().rev() {
            let curr = value_map[&ch];
            if curr < last {
                ans -= curr;
            } else {
                ans += curr;
            }
            last = curr
        }
        ans
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s = String::from("III");
        let ans = 3;
        assert_eq!(Solution::roman_to_int(s), ans);
    }

    #[test]
    fn example2() {
        let s = String::from("IV");
        let ans = 4;
        assert_eq!(Solution::roman_to_int(s), ans);
    }

    #[test]
    fn example3() {
        let s = String::from("IX");
        let ans = 9;
        assert_eq!(Solution::roman_to_int(s), ans);
    }

    #[test]
    fn example4() {
        let s = String::from("LVIII");
        let ans = 58;
        assert_eq!(Solution::roman_to_int(s), ans);
    }

    #[test]
    fn example5() {
        let s = String::from("MCMXCIV");
        let ans = 1994;
        assert_eq!(Solution::roman_to_int(s), ans);
    }
}
